File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/HiddenModelCreator.java
package org.aiddl.example.learning_agent.functions;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.aiddl.common.java.learning.LearningTerm;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.CollectionTerm;
import org.aiddl.core.java.representation.Term;

public class HiddenModelCreator implements Function {

	Random rng = new Random();
	
	@Override
	public Term apply( Term args ) {
		CollectionTerm locations = (CollectionTerm) args.get(Term.sym("locations"));
		CollectionTerm configs = (CollectionTerm) args.get(Term.sym("configs"));
		Term label = args.get(LearningTerm.Label);
		
		List<Term> map = new ArrayList<>();
		for ( Term l : locations ) {
			for ( Term c : configs ) {
				map.add(Term.keyVal(Term.tuple(l, c), label.get(rng.nextInt(label.size()))));
			}
		}

		return Term.list(map);
	}
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/HiddenModelCreator.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{CollectionTerm, KeyVal, ListTerm, Sym, Term, Tuple}

import scala.util.Random

class HiddenModelCreator extends Function {

  private val rng = new Random()

  override def apply(x: Term): Term = {
    val locations = x(Sym("locations")).asCol
    val configs = x(Sym("configs")).asCol
    val label = x(Sym("label")).asList
    this(locations, configs, label)
  }

  def apply(locations: CollectionTerm, configs: CollectionTerm, label: ListTerm): ListTerm = {
    var list: List[Term] = Nil

    for (l <- locations) {
      for (c <- configs) {
        list = KeyVal(
          Tuple(l, c),
          label(rng.nextInt(label.size))
        ) :: list
      }
    }
    ListTerm(list)
  }

}
